builder/build_validate.go

```diff
@@ -254,46 +263,82 @@ func (b *Builder) mcaPkgInfo(mInfo map[string]*swupd.Manifest, version, download
 	pInfo := make(map[string]*mcaBundlePkgInfo)
 
 	// Download and query file metadata from all packages in each bundle
+	bundleWorker := func() {
+		for m := range mCh {
+			var pkgList = []string{}
+			for pkg := range mInfo[m.name].BundleInfo.AllPackages {
+				pkgList = append(pkgList, pkg)
+			}
+
+			out, err := downloadRpms(packagerCmd, pkgList, buildVersionDir, downloadRetries)
+			if err != nil {
+				errorCh <- err
+				wg.Done()
+				return
+			}
+
+			// Collect metadata to resolve installed RPM file names
+			bundlePkgInfo := pkgInfoFromNoopInstall(out.String())
+			for _, p := range bundlePkgInfo {
+				// Resolve files for package when it doesn't exist in the cache
+				rw.RLock()
```

Comment:
instead of locking the entire if block, we could do:
```
rw.RLock()
name := pkgCache[p.name]
rw.RUnlock()

if name == nil {
...
} else {
    m.allPkgs[p.name] = name
}
```

This way `resolvePkgFiles()` will always run outside the lock context and won't prevent an write lock to be acquired if  this goroutine is stopped when exec() is called.

Reply:
Good idea! I pushed the change.